pkg/collector/reconcile/servicemonitor.go

```diff
@@ -0,0 +1,170 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconcile
+
+import (
+	"context"
+	"fmt"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/open-telemetry/opentelemetry-operator/pkg/featuregate"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/naming"
+)
+
+// +kubebuilder:rbac:groups=monitoring.coreos.com,resources=servicemonitors,verbs=get;list;watch;create;update;patch;delete
+
+// ServiceMonitors reconciles the service monitor(s) required for the instance in the current context.
+func ServiceMonitors(ctx context.Context, params Params) error {
+	if !params.Instance.Spec.Observability.Metrics.CreateServiceMonitors || !featuregate.PrometheusOperatorIsAvailable.IsEnabled() {
+		return nil
+	}
+
+	desired := desiredServiceMonitors(ctx, params)
+
+	// first, handle the create/update parts
+	if err := expectedServiceMonitors(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the expected service monitors: %w", err)
+	}
+
+	// then, delete the extra objects
+	if err := deleteServiceMonitors(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the service monitors to be deleted: %w", err)
+	}
+
+	return nil
+}
+
+func desiredServiceMonitors(_ context.Context, params Params) []monitoringv1.ServiceMonitor {
+	col := params.Instance
+	return []monitoringv1.ServiceMonitor{
```

Comment:
I don't love the idea of the OpenTelemetry Operator being able to create objects that depend on other CRDs being preset, it feels like an opportunity for our users to get into weird situations.

Reply:
this also creates a direct dependency on the prometheus operator from ours which concerns me. Right now we have a dependency in the TA which makes sense as it is the component pulling down that functionality, but I worry what happens with this added dependency.

Reply:
> I don't love the idea of the OpenTelemetry Operator being able to create objects that depend on other CRDs being preset

From the conversations we had during the 25 May SIG meeting, I understood it was OK to implement the proposal described in #1768.

> it feels like an opportunity for our users to get into weird situations.

What situations? I would be happy to check them and apply a fix if needed.

> I agree, I think this opens up some issues if we make the assumption that every operator user wants to use service monitors. 

I don't think this will be a problem since everything is protected with a feature flag. If the feature flag is not enabled, the Service Monitors will not be created. Also, you need to set to true the feature in the OpenTelemetry Collector instance.

> This is a good topic for our next SIG meeting.

I'll not be able to attend since I'll be on vacation next week. But I'll be happy to talk about it in another moment.

Reply:
We have discussed this on the SIG meeting that the operator will create Service/Pod monitors for operands with the assumption that the functionality will be put behind a feature flag. The implementation has to be flexible to allow OTLP in the future (once the collector will report OTLP metrics).

This work is needed to move the OTEL operator to operator level 4 https://operatorhub.io/operator/opentelemetry-operator. There is a lot of other operators that do the same.

Reply:
ah okay, yeah sorry checked my notes and saw this now. Proceed :) will review now.